tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,644 @@
+    ///
+    /// # fn main() -> std::io::Result<()> {
+    /// let builder = NamedPipeClientBuilder::new("\\\\.\\pipe\\mynamedpipe");
+    /// let pipe = builder.create()?;
+    /// # Ok(()) }
+    /// ```
+    pub fn new(addr: impl AsRef<OsStr>) -> Self {
+        Self {
+            name: addr
+                .as_ref()
+                .encode_wide()
+                .chain(Some(0))
+                .collect::<Arc<[u16]>>(),
+        }
+    }
+
+    /// Waits until either a time-out interval elapses or an instance of the
+    /// specified named pipe is available for connection (that is, the pipe's
+    /// server process has a pending [connect] operation on the pipe).
+    ///
+    /// This corresponds to the [`WaitNamedPipeW`] system call.
+    ///
+    /// [`WaitNamedPipeW`]:
+    /// https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-waitnamedpipea
+    /// [connect]: NamedPipe::connect
+    ///
+    /// ```no_run
+    /// use tokio::net::windows::NamedPipeClientBuilder;
+    ///
+    /// # #[tokio::main] async fn main() -> std::io::Result<()> {
+    /// let builder = NamedPipeClientBuilder::new("\\\\.\\pipe\\mynamedpipe");
+    ///
+    /// // Wait forever until a socket is available to be connected to.
+    /// builder.wait(None).await?;
+    ///
+    /// let pipe = builder.create()?;
+    /// # Ok(()) }
+    /// ```
+    ///
+    /// # Panics
+    ///
+    /// Panics if the specified duration is larger than `0xffffffff`
+    /// milliseconds, which is roughly equal to 1193 hours.
+    ///
+    /// ```should_panic
+    /// use std::time::Duration;
+    /// use tokio::net::windows::NamedPipeClientBuilder;
+    ///
+    /// # #[tokio::main] async fn main() -> std::io::Result<()> {
+    /// let builder = NamedPipeClientBuilder::new("\\\\.\\pipe\\mynamedpipe");
+    ///
+    /// builder.wait(Some(Duration::from_millis(0xffffffff))).await?;
+    /// # Ok(()) }
+    /// ```
+    pub async fn wait(&self, timeout: Option<Duration>) -> io::Result<()> {
+        let timeout = match timeout {
+            Some(timeout) => {
+                let timeout = timeout.as_millis();
+                assert! {
+                    timeout < NMPWAIT_WAIT_FOREVER as u128,
+                    "timeout out of bounds, can wait at most {}ms, but got {}ms", NMPWAIT_WAIT_FOREVER - 1,
+                    timeout
+                };
+                timeout as DWORD
+            }
+            None => NMPWAIT_WAIT_FOREVER,
+        };
+
+        let name = self.name.clone();
+
+        // TODO: Is this the right thread pool to use? `WaitNamedPipeW` could
+        // potentially block for a fairly long time all though it's only
+        // expected to be used when connecting something which should be fairly
+        // constrained.
+        //
+        // But doing something silly like spawning hundreds of clients trying to
+        // connect to a named pipe server without a timeout could easily end up
+        // starving the thread pool.
```

Comment:
Yeah, we might want to start out without this feature. People who need it can still do it using `spawn_blocking` themselves. It's especially problematic that the spawned task cannot be cancelled.

Reply:
Ugh, yeah.

The benefits of this are:
* It allows for calling `WaitNamedPipe` without cloning the whole builder.
* The `spawn_blocking` dance is really awkward 😭

I'm gonna dig around for alternatives but I'll consider this a blocker before merging.

Reply:
It is still included, but no longer async.